File: ios/App/GolfWatch Watch App/Practice/SwingMetricsCalculator.swift
import Foundation
import simd

/// Turns a completed `SwingWindow` into compact `SwingMetrics`. All outputs
/// are relative/estimated — timings in ms plus 0-100 scores. Nothing here is
/// an absolute physical measurement of the club or ball.
struct SwingMetricsCalculator {
    func compute(_ w: SwingWindow) -> SwingMetrics {
        let backswingMs = (w.tTop - w.tStart) * 1000
        let downswingMs = (w.tImpact - w.tTop) * 1000
        let tempo = downswingMs > 1 ? backswingMs / downswingMs : 0

        let peakOmega = w.samples.map(\.angularSpeed).max() ?? 0
        let backswingRotation = integratedRotation(w, from: w.tStart, to: w.tTop)

        return SwingMetrics(
            backswingTimeMs: Int(backswingMs.rounded()),
            downswingTimeMs: Int(downswingMs.rounded()),
            tempoRatio: round2(tempo),
            transitionScore: transitionScore(w),
            estimatedHandSpeed: handSpeedScore(w),
            wristRotationScore: wristRotationScore(w),
            finishStabilityScore: finishStabilityScore(w),
            planeAxis: meanRotationAxis(w, from: w.tTop, to: w.tImpact),
            swingType: classifySwing(w, peakOmega: peakOmega, backswingRotation: backswingRotation),
            isAirSwing: w.isAirSwing,
            backswingRotation: round2(backswingRotation),
            releaseTimingScore: releaseTimingScore(w),
            decelerationScore: decelerationScore(w),
            transitionDirectionScore: transitionDirectionScore(w),
            addressGravity: addressGravity(w),
            peakImpactG: w.peakImpactG
        )
    }

    // MARK: - Phase 1 derived signals

    /// Estimated swing classification from motion magnitude + duration +
    /// whether contact occurred. Thresholds are rough — labeled "estimated".
    private func classifySwing(_ w: SwingWindow, peakOmega: Double, backswingRotation: Double) -> String {
        if w.isAirSwing { return "air" }
        let totalMs = (w.tFinish - w.tStart) * 1000
        // Putts: very low wrist speed and a short, quiet stroke.
        if peakOmega < 5 && backswingRotation < 1.2 { return "putt" }
        // Chips: low speed, short backswing.
        if peakOmega < 9 || backswingRotation < 1.6 { return "chip" }
        // Pitches: moderate speed / partial backswing or a shorter total.
        if peakOmega < 16 || backswingRotation < 2.4 || totalMs < 900 { return "pitch" }
        return "full"
    }

    /// Integrate |angular velocity| over a sub-window → total wrist rotation
    /// (radians). A motion amount; the UI shows it as relative short/normal/long.
    private func integratedRotation(_ w: SwingWindow, from: TimeInterval, to: TimeInterval) -> Double {
        let seg = w.samples.filter { $0.t >= from && $0.t <= to }
        guard seg.count > 1 else { return 0 }
        var total = 0.0
        for i in 1..<seg.count {
            let dt = seg[i].t - seg[i - 1].t
            total += seg[i].angularSpeed * dt
        }
        return total
    }

    /// 0-100: how late in the downswing the wrist's peak speed occurs. A later
    /// peak (closer to impact) reads as more retained speed / lag.
    private func releaseTimingScore(_ w: SwingWindow) -> Int {
        let down = w.samples.filter { $0.t >= w.tTop && $0.t <= w.tImpact }
        guard down.count > 1, w.tImpact > w.tTop else { return 50 }
        let peak = down.max(by: { $0.angularSpeed < $1.angularSpeed })
        guard let tPeak = peak?.t else { return 50 }
        let frac = (tPeak - w.tTop) / (w.tImpact - w.tTop)
        return clampScore(frac * 100)
    }

    /// 0-100: accelerating through impact (high) vs quitting on it (low).
    /// Compares wrist speed in the last ~50ms before impact to the ~50ms before
    /// that.
    private func decelerationScore(_ w: SwingWindow) -> Int {
        let late = w.samples.filter { $0.t >= w.tImpact - 0.05 && $0.t <= w.tImpact }
        let early = w.samples.filter { $0.t >= w.tImpact - 0.10 && $0.t < w.tImpact - 0.05 }
        guard !late.isEmpty, !early.isEmpty else { return 50 }
        let lateAvg = mean(late.map(\.angularSpeed))
        let earlyAvg = mean(early.map(\.angularSpeed))
        guard earlyAvg > 1e-3 else { return 50 }
        // Still speeding up → >50; slowing down → <50.
        let ratio = (lateAvg - earlyAvg) / earlyAvg
        return clampScore(50 + ratio * 100)
    }

    /// 0-100: how consistent the rotation axis stays from backswing to
    /// downswing. A large shift reads as an "over-the-top"-style direction
    /// change. A motion tendency — NOT a measured club path.
    private func transitionDirectionScore(_ w: SwingWindow) -> Int {
        let backAxis = meanRotationAxis(w, from: w.tStart, to: w.tTop)
        let downAxis = meanRotationAxis(w, from: w.tTop, to: w.tImpact)
        let sim = dot(backAxis, downAxis) // -1..1, 1 = same direction
        return clampScore((sim + 1) / 2 * 100)
    }

    /// Gravity direction at takeaway — a proxy for setup/address orientation,
    /// compared across swings on the phone for setup repeatability.
    private func addressGravity(_ w: SwingWindow) -> [Double] {
        guard let first = w.samples.first else { return [0, 0, 0] }
        let g = first.gravity
        return [g.x, g.y, g.z]
    }

    /// Smoothness of the velocity reversal at the top: lower jerk (rate of
    /// change of acceleration) through a window around the top = smoother.
    private func transitionScore(_ w: SwingWindow, windowS: Double = 0.12) -> Int {
        let lo = w.tTop - windowS, hi = w.tTop + windowS
        let seg = w.samples.filter { $0.t >= lo && $0.t <= hi }
        guard seg.count > 2 else { return 50 }
        var jerk = 0.0
        for i in 1..<seg.count {
            let dt = max(seg[i].t - seg[i - 1].t, 1e-3)
            jerk += simd_length(seg[i].userAccel - seg[i - 1].userAccel) / dt
        }
        let avgJerk = jerk / Double(seg.count - 1)
        // 60 g/s ≈ an aggressive transition; clamp to a 0-100 smoothness score.
        return clampScore((1 - min(avgJerk / 60.0, 1)) * 100)
    }

    /// Peak angular speed near impact mapped to a RELATIVE 0-100 effort scale.
    /// Deliberately not converted to mph or club-head speed.
    private func handSpeedScore(_ w: SwingWindow) -> Int {
        let peak = w.samples.map(\.angularSpeed).max() ?? 0
        return clampScore(peak / 35.0 * 100) // 35 rad/s ≈ a fast wrist
    }

    /// Consistency/smoothness of forearm roll through the downswing.
    private func wristRotationScore(_ w: SwingWindow) -> Int {
        let down = w.samples.filter { $0.t >= w.tTop && $0.t <= w.tImpact }
        guard !down.isEmpty else { return 50 }
        let rolls = down.map { abs($0.rotationRate.z) } // z ≈ crown/forearm axis
        let m = mean(rolls)
        let varc = mean(rolls.map { ($0 - m) * ($0 - m) })
        return clampScore((1 - min(varc / (m * m + 1e-3), 1)) * 100)
    }

    /// Inverse of acceleration variance in the ~600ms after impact: steadier
    /// = more stable finish.
    private func finishStabilityScore(_ w: SwingWindow) -> Int {
        let lo = w.tImpact, hi = w.tImpact + 0.6
        let seg = w.samples.filter { $0.t >= lo && $0.t <= hi }
        guard !seg.isEmpty else { return 50 }
        let mags = seg.map(\.linearAccelMag)
        let m = mean(mags)
        let varc = mean(mags.map { ($0 - m) * ($0 - m) })
        return clampScore((1 - min(varc / 1.5, 1)) * 100)
    }

    /// Mean angular-velocity direction over a sub-window, normalised. Used only
    /// to compare swing-motion-pattern consistency across swings — it is NOT an
    /// absolute swing-plane angle.
    private func meanRotationAxis(_ w: SwingWindow, from: TimeInterval, to: TimeInterval) -> [Double] {
        let seg = w.samples.filter { $0.t >= from && $0.t <= to }
        var acc = simd_double3(0, 0, 0)
        for s in seg { acc += s.rotationRate }
        let n = simd_length(acc)
        let unit = n > 1e-6 ? acc / n : simd_double3(0, 0, 0)
        return [unit.x, unit.y, unit.z]
    }

    // --- helpers ---
    private func mean(_ xs: [Double]) -> Double {
        xs.isEmpty ? 0 : xs.reduce(0, +) / Double(xs.count)
    }
    private func dot(_ a: [Double], _ b: [Double]) -> Double {
        var s = 0.0
        for i in 0..<min(a.count, b.count) { s += a[i] * b[i] }
        return s
    }
    private func clampScore(_ v: Double) -> Int { max(0, min(100, Int(v.rounded()))) }
    private func round2(_ v: Double) -> Double { (v * 100).rounded() / 100 }
}

File: android/wear/src/main/java/com/golfroundtracker/app/wear/SwingMetricsCalculator.kt
package com.golfroundtracker.app.wear

import kotlin.math.abs
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sqrt

/**
 * Turns a completed `SwingWindow` into `SwingMetrics`.
 *
 * A faithful port of `SwingMetricsCalculator.swift`. Every threshold here (35
 * rad/s for a fast wrist, 60 g/s for an aggressive transition, the swing-type
 * cutoffs) was chosen against real swings, so the numbers are copied rather
 * than re-derived — the phone's rules engine and the `swing_metrics` table are
 * calibrated to THESE scales, and a Wear watch producing subtly different ones
 * would poison a shared history without ever looking wrong.
 */
class SwingMetricsCalculator {

    fun compute(w: SwingWindow): SwingMetrics {
        val backswingMs = (w.tTop - w.tStart) * 1000
        val downswingMs = (w.tImpact - w.tTop) * 1000
        val tempo = if (downswingMs > 1) backswingMs / downswingMs else 0.0

        val peakOmega = w.samples.maxOfOrNull { it.angularSpeed } ?: 0.0
        val backswingRotation = integratedRotation(w, w.tStart, w.tTop)

        return SwingMetrics(
            backswingTimeMs = backswingMs.roundToInt(),
            downswingTimeMs = downswingMs.roundToInt(),
            tempoRatio = round2(tempo),
            transitionScore = transitionScore(w),
            estimatedHandSpeed = handSpeedScore(w),
            wristRotationScore = wristRotationScore(w),
            finishStabilityScore = finishStabilityScore(w),
            planeAxis = meanRotationAxis(w, w.tTop, w.tImpact),
            swingType = classifySwing(w, peakOmega, backswingRotation),
            isAirSwing = w.isAirSwing,
            backswingRotation = round2(backswingRotation),
            releaseTimingScore = releaseTimingScore(w),
            decelerationScore = decelerationScore(w),
            transitionDirectionScore = transitionDirectionScore(w),
            addressGravity = addressGravity(w),
            peakImpactG = w.peakImpactG
        )
    }

    /** Motion magnitude + duration + whether contact happened. Rough by design,
     *  and labelled "estimated" everywhere it surfaces. */
    private fun classifySwing(w: SwingWindow, peakOmega: Double, backswingRotation: Double): String {
        if (w.isAirSwing) return "air"
        val totalMs = (w.tFinish - w.tStart) * 1000
        if (peakOmega < 5 && backswingRotation < 1.2) return "putt"
        if (peakOmega < 9 || backswingRotation < 1.6) return "chip"
        if (peakOmega < 16 || backswingRotation < 2.4 || totalMs < 900) return "pitch"
        return "full"
    }

    /** Integrate |angular velocity| over a sub-window → radians of wrist turn. */
    private fun integratedRotation(w: SwingWindow, from: Double, to: Double): Double {
        val seg = w.samples.filter { it.t in from..to }
        if (seg.size <= 1) return 0.0
        var total = 0.0
        for (i in 1 until seg.size) {
            total += seg[i].angularSpeed * (seg[i].t - seg[i - 1].t)
        }
        return total
    }

    /** 0-100: how late peak wrist speed lands in the downswing — lag-like. */
    private fun releaseTimingScore(w: SwingWindow): Int {
        val down = w.samples.filter { it.t in w.tTop..w.tImpact }
        if (down.size <= 1 || w.tImpact <= w.tTop) return 50
        val tPeak = down.maxByOrNull { it.angularSpeed }?.t ?: return 50
        return clampScore((tPeak - w.tTop) / (w.tImpact - w.tTop) * 100)
    }

    /** 0-100: accelerating through impact (>50) vs quitting on it (<50). */
    private fun decelerationScore(w: SwingWindow): Int {
        val late = w.samples.filter { it.t >= w.tImpact - 0.05 && it.t <= w.tImpact }
        val early = w.samples.filter { it.t >= w.tImpact - 0.10 && it.t < w.tImpact - 0.05 }
        if (late.isEmpty() || early.isEmpty()) return 50
        val lateAvg = late.map { it.angularSpeed }.mean()
        val earlyAvg = early.map { it.angularSpeed }.mean()
        if (earlyAvg <= 1e-3) return 50
        return clampScore(50 + (lateAvg - earlyAvg) / earlyAvg * 100)
    }

    /** 0-100: axis consistency backswing → downswing. An over-the-top PROXY,
     *  not a measured club path. */
    private fun transitionDirectionScore(w: SwingWindow): Int {
        val backAxis = meanRotationAxis(w, w.tStart, w.tTop)
        val downAxis = meanRotationAxis(w, w.tTop, w.tImpact)
        val sim = dot(backAxis, downAxis) // -1..1
        return clampScore((sim + 1) / 2 * 100)
    }

    private fun addressGravity(w: SwingWindow): List<Double> {
        val first = w.samples.firstOrNull() ?: return listOf(0.0, 0.0, 0.0)
        return listOf(first.gravX, first.gravY, first.gravZ)
    }

    /** Smoothness of the reversal at the top — lower jerk is smoother. */
    private fun transitionScore(w: SwingWindow, windowS: Double = 0.12): Int {
        val seg = w.samples.filter { it.t in (w.tTop - windowS)..(w.tTop + windowS) }
        if (seg.size <= 2) return 50
        var jerk = 0.0
        for (i in 1 until seg.size) {
            val dt = maxOf(seg[i].t - seg[i - 1].t, 1e-3)
            val dx = seg[i].ax - seg[i - 1].ax
            val dy = seg[i].ay - seg[i - 1].ay
            val dz = seg[i].az - seg[i - 1].az
            jerk += sqrt(dx * dx + dy * dy + dz * dz) / dt
        }
        val avgJerk = jerk / (seg.size - 1)
        // 60 g/s ≈ an aggressive transition.
        return clampScore((1 - min(avgJerk / 60.0, 1.0)) * 100)
    }

    /** Peak angular speed as RELATIVE 0-100 effort. Never mph. */
    private fun handSpeedScore(w: SwingWindow): Int {
        val peak = w.samples.maxOfOrNull { it.angularSpeed } ?: 0.0
        return clampScore(peak / 35.0 * 100) // 35 rad/s ≈ a fast wrist
    }

    /** Consistency of forearm roll through the downswing. */
    private fun wristRotationScore(w: SwingWindow): Int {
        val down = w.samples.filter { it.t in w.tTop..w.tImpact }
        if (down.isEmpty()) return 50
        // z ≈ the crown/forearm axis, the same assumption the Swift version
        // makes — and the same one that would need revisiting if a Wear device
        // turns out to orient its gyroscope differently on the wrist.
        val rolls = down.map { abs(it.gz) }
        val m = rolls.mean()
        val varc = rolls.map { (it - m) * (it - m) }.mean()
        return clampScore((1 - min(varc / (m * m + 1e-3), 1.0)) * 100)
    }

    /** Steadiness in the ~600 ms after impact. */
    private fun finishStabilityScore(w: SwingWindow): Int {
        val seg = w.samples.filter { it.t in w.tImpact..(w.tImpact + 0.6) }
        if (seg.isEmpty()) return 50
        val mags = seg.map { it.linearAccelMag }
        val m = mags.mean()
        val varc = mags.map { (it - m) * (it - m) }.mean()
        return clampScore((1 - min(varc / 1.5, 1.0)) * 100)
    }

    /** Mean angular-velocity direction, normalised. For comparing PATTERNS. */
    private fun meanRotationAxis(w: SwingWindow, from: Double, to: Double): List<Double> {
        val seg = w.samples.filter { it.t in from..to }
        var x = 0.0; var y = 0.0; var z = 0.0
        for (s in seg) { x += s.gx; y += s.gy; z += s.gz }
        val n = sqrt(x * x + y * y + z * z)
        return if (n > 1e-6) listOf(x / n, y / n, z / n) else listOf(0.0, 0.0, 0.0)
    }

    private fun List<Double>.mean(): Double = if (isEmpty()) 0.0 else sum() / size

    private fun dot(a: List<Double>, b: List<Double>): Double {
        var s = 0.0
        for (i in 0 until min(a.size, b.size)) s += a[i] * b[i]
        return s
    }

    private fun clampScore(v: Double): Int = v.roundToInt().coerceIn(0, 100)

    private fun round2(v: Double): Double = (v * 100).roundToInt() / 100.0
}
